### PR TITLE
fix(projects): 项目 clone 失败后任务仍可启动的事故链收口

### PR DESCRIPTION
**Pull request by lvhaojie456**
## 问题（线上事故链）

创建 Project 后 clone 在后台异步执行、接口立即返回成功；HTTPS 无凭据时后台 clone 因无 TTY 报 `could not read Username ... No such device or address`，项目标 `error` 但 `local_path` 保留指向不存在的目录。Task 创建与 dispatcher 领取都不检查 `project.status`，dispatcher 以 `last_cwd or target_repo` 直接作 cwd；Codex `thread/start` 的三个 required MCP 报 `No such file or directory` 后走 safe-fallback，`create_subprocess_exec(cwd=...)` 抛 `FileNotFoundError`，页面最终只显示缺上下文的裸 `[Errno 2]`，且每次失败烧掉 retry 预算（3 次后永久 failed，Re-clone 也救不回）。

## 修复（三层防御）

1. **出队门禁**：`task_queue.project_ready_dispatch_predicate`（否定式 correlated-EXISTS）同时加进 dequeue 候选 SELECT 与领取 CAS。非 ready 项目的 Task 留在 pending（`project_id` 为 NULL 或悬空不受影响、不写 Task.status、不烧 retry 预算），clone/Re-clone 成功回 ready 后 wake 自动开跑（2s poll 兜底）。
2. **入口拒绝**：POST /api/tasks（乐观+锁内两处）、PUT 换 project、Todo Run、standalone Plan 物化经新增 `project_readiness.require_project_dispatchable` 对 `error` 项目返回 422（migration-import / shared shadow / chat fork / PR Review 内部 / Harness child 豁免，故未放进 `stage_task_record`）。前端 ProjectSelect 对 error 项目红色标注+悬停原因，TaskForm/PlanCreateForm 显示警示条并禁用提交，cloning 项目提示等待。
3. **启动兜底**：新增 `TaskWorkingDirectoryMissingError`（继承 `TaskAgentIsolationError`）：`prepare_task_working_directory` 对显式 workspace 校验存在性，`_spawn_managed_direct_process` 把 cwd 缺失的 ENOENT 翻译成同一人话异常（可执行文件缺失保持原样）。queued chat 将其并入 permanent prelaunch（聊天红色通知 + Task 回滚发送前状态 + 消息不重排，替代旧的无限静默重排）；Monitor/Sub-Agent 直 spawn 路径经 `require_existing_task_cwd` preflight（Monitor 保留有界退避，Sub-Agent 失败补写 `last_error`）。

## 附带修复

- 后台 clone 无条件 `GIT_TERMINAL_PROMPT=0` + `stdin=DEVNULL`、SSH 追加 `BatchMode=yes`（仅 clone 作用域，未动公共 `_build_git_env`）；认证类 stderr 归一化为可操作文案
- clone 失败/成功由 `_sync_waiting_task_clone_notes` 批量注记/清除等待任务的 `error_message`（不动 status，无广播义务）
- `worker_proxy.ensure_worker_project` 对远端 clone `error` 秒级失败（原来白等 300s 超时）
- **Codex runtime scope 竞态**：Codex launch 补齐与 Claude 对等的 task runtime scope reserve → exact-generation adopt（app-server adapter 与 exec 进程都在注册的同一同步步骤 adopt），dispatcher finally 保持原样、自动变成无 owner 兜底，闭合「旧 lifecycle finally 删掉新代次已物化、MCP 子进程尚未 spawn 的入口文件」的窄窗口竞态，并收口 Codex chat turn 的 scope 泄漏

## 用户可感知的行为变更

1. error 项目不能再创建任务（422），需先 Re-clone
2. chat 消息在工作目录缺失时从「无限静默重试」变为「明确拒绝、修复后重发」
3. Codex turn 结束后 runtime scope 立即回收（原来泄漏到下个 fresh turn）

## 验证

- Linux 容器（python:3.12 + `--init`，干净 clone + patch）：新增 `test_project_readiness.py` 13 个测试全绿（含红绿验证）；6 个相关测试文件定向回归 **520 passed / 2 skipped**，唯一失败（`test_delete_task_preserves_reaped_parent_with_live_generation`）在未打补丁的基线上以相同方式失败（环境性），零回归
- 前端 `npx tsc --noEmit` 零错误；TaskForm/PlanCreateForm vitest 48 passed
- 文档已同步：CLAUDE.md 关键约定、TEST.md 用例清单、PROGRESS.md 复盘、中英 README

⚠️ 本次开发环境为 Windows，`fcntl` 依赖的用例已按平台 skip、Linux 上全量执行；建议合入前在 Linux 上补跑一次后端整库回归。

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by youchengsong — COMMENTED** [review]
# CCM reviewer panel: changes required

The panel found 7 open blocking findings on this reviewed PR head. Address the inline finding threads before merging.

## Reviewer summaries

### Principal engineer — Changes required

The readiness gate is directionally sound, but SSH clones are not consistently noninteractive and successful clones can be dispatched before later initialization reverses the Project to error.

Finding count: 2 total (2 blocking, 0 advisory).

### Senior engineer — Changes required

The readiness gate is broadly coherent, but the patch leaves SSH clones interactive, can wake tasks before clone-related error handling is final, and over-applies update admission to unchanged project assignments.

Finding count: 3 total (3 blocking, 0 advisory).

### QA engineer — Changes required

The readiness gating is broadly covered, but the new Codex runtime-scope reservation can leak on pre-launch exceptions and the claimed concurrency fix lacks regression coverage.

Finding count: 2 total (2 blocking, 0 advisory).

Detailed evidence, impact, required fixes, and verification steps are published in dedicated finding threads/comments.

File: backend/api/projects.py [hunk 1]
```diff
@@ -928,7 +1002,15 @@ async def _clone_repo(project_id: int, git_url: str, local_path: str, project_na
 
         # Build env with git credentials so clone/fetch can authenticate
         git_env = _build_git_env(git_config or {})
-        env = {**os.environ, **git_env} if git_env else None
+        env = {**os.environ, **git_env}
+        # Background clones have no TTY to answer a credential prompt: git
+        # either hangs the dev shell or fails with the cryptic "could not read
+        # Username ... No such device or address" under systemd. Fail fast
```
Comment (youchengsong) [inline backend/api/projects.py:1021]: **[MEDIUM] SSH clones remain interactive when no SSH command is configured**

Reviewer: `principal_engineer` · Category: `operations`

Evidence: The clone path adds `BatchMode=yes` only when `GIT_SSH_COMMAND` already exists. For the normal SSH URL case with no credential-specific or host-level `GIT_SSH_COMMAND`, only `GIT_TERMINAL_PROMPT=0` and `stdin=DEVNULL` are applied; the former controls Git credential prompting, not SSH authentication or host-key prompting. The updated test explicitly permits `GIT_SSH_COMMAND` to be absent.

Impact: An SSH clone can still invoke SSH's interactive authentication or host-key flow, contradicting the fail-fast guarantee and potentially leaving the background clone waiting or producing inconsistent, unactionable failures.

Required fix: For every background clone, construct a clone-scoped SSH command containing `-o BatchMode=yes`, using the configured/environment command when present and a safe `ssh` default otherwise. Do not change the shared `_build_git_env`.

Verification: Add a clone test using an SSH URL with no git configuration and no inherited `GIT_SSH_COMMAND`; assert the subprocess environment contains an SSH command with `BatchMode=yes`, alongside `GIT_TERMINAL_PROMPT=0` and `stdin=DEVNULL`.

Finding fingerprint: `12c3db55ad209fe7d32f66ebf34665c3857ec99a0cf06de4e7a4bb67d72680f3`

File: backend/api/projects.py [hunk 2]
```diff
@@ -983,6 +1066,12 @@ async def _clone_repo(project_id: int, git_url: str, local_path: str, project_na
             )
             await db.commit()
 
+        # Tasks created while the clone was still running are held back by the
```
Comment (youchengsong) [inline backend/api/projects.py]: **[MEDIUM] Dispatcher is woken before clone success is final**

Reviewer: `principal_engineer` · Category: `concurrency`

Evidence: After committing the Project as ready, `_clone_repo` awaits note synchronization and calls `_wake_dispatcher()`, then runs manager-side `try_auto_configure_pr_monitor` and `try_auto_configure_delivery_monitor`. Those later awaits remain inside the outer `try`; any exception reaches the outer handler, which changes the same Project to `error` and annotates pending tasks. The wake can therefore claim a task while those operations are running, before the readiness state is potentially reversed.

Impact: A task associated with a Project that ultimately ends in `error` may already be in progress, violating the new invariant that non-ready Projects cannot dispatch and recreating the launch failure/retry-budget incident under a post-clone setup failure.

Required fix: Make the ready transition and dispatcher wake the final successful state publication: perform all failure-producing initialization before publishing `ready`, or isolate optional post-clone setup failures so they cannot reverse clone readiness. Wake only after the authoritative final state is committed.

Verification: Block the manager auto-configuration coroutine after clone completion, verify dequeue cannot claim the queued task before final readiness publication, then make auto-configuration fail and assert the Project is error and the task remains pending and unclaimed. Also verify the success path publishes ready and wakes/dispatches afterward.

Finding fingerprint: `3eab257fdbc17efda4b865266c2527733af9363964f6eca58b9a3a596e487eef`

File: backend/api/projects.py [hunk 3]
```diff
@@ -928,7 +1002,15 @@ async def _clone_repo(project_id: int, git_url: str, local_path: str, project_na
...
         # Build env with git credentials so clone/fetch can authenticate
         git_env = _build_git_env(git_config or {})
-        env = {**os.environ, **git_env} if git_env else None
+        env = {**os.environ, **git_env}
+        # Background clones have no TTY to answer a credential prompt: git
+        # either hangs the dev shell or fails with the cryptic "could not read
+        # Username ... No such device or address" under systemd. Fail fast
+        # instead — GIT_ASKPASS (set when a token is configured) is consulted
+        # before terminal prompts, so authenticated clones are unaffected.
+        env["GIT_TERMINAL_PROMPT"] = "0"
```
Comment (youchengsong) [inline backend/api/projects.py:1024]: **[MEDIUM] SSH batch mode is omitted when no SSH command is configured**

Reviewer: `senior_engineer` · Category: `operations`

Evidence: The clone environment appends `-o BatchMode=yes` only under `if "GIT_SSH_COMMAND" in env`. A normal SSH clone with neither host nor Git configuration defining that variable therefore receives no batch-mode option. The added test explicitly permits this case with `"GIT_SSH_COMMAND" not in clone_env or ...`, so it does not prove the claimed non-interactive behavior.

Impact: An SSH clone can still attempt interactive host-key, password, or key-passphrase prompting. SSH may open `/dev/tty` independently of the subprocess stdin, so `stdin=DEVNULL` and `GIT_TERMINAL_PROMPT=0` do not fully prevent a background clone from hanging.

Required fix: Always provide a batch-mode SSH command for clone operations, using the configured/inherited command when present and an `ssh` default otherwise, while preserving any existing options.

Verification: Run `_clone_repo` with an SSH URL and no `GIT_SSH_COMMAND`; assert the subprocess environment contains an SSH command with `BatchMode=yes`. Also cover augmentation of an existing command.

Finding fingerprint: `e35734930d5fedb4e77c65bd210299aebf349d811cd913c0b0b34d6d6d241431`

File: backend/api/projects.py [hunk 4]
```diff
@@ -983,6 +1066,12 @@ async def _clone_repo(project_id: int, git_url: str, local_path: str, project_na
             )
             await db.commit()
 
+        # Tasks created while the clone was still running are held back by the
+        # dispatch-queue readiness gate; nudge the dispatcher so they start now
+        # (the 2s poll is only the fallback) and drop any stale failure note.
+        await _sync_waiting_task_clone_notes(project_id, None)
```
Comment (youchengsong) [inline backend/api/projects.py]: **[MEDIUM] Dispatcher is woken before the clone success path can no longer become an error**

Reviewer: `senior_engineer` · Category: `concurrency`

Evidence: After committing the Project as ready, the patch clears task notes and calls `_wake_dispatcher()`, then awaits manager-only GitHub setup. Those later calls remain inside the outer `try`; any exception reaches the outer `except` and changes the same Project to `error`. The wake allows a waiting task to be claimed between those two state transitions.

Impact: A task associated with a Project that ultimately ends this coroutine in `error` can enter execution, defeating the new queue invariant and potentially consuming work or retry budget under a state that should remain gated.

Required fix: Make the ready transition and wake the final successful publication step, or isolate optional post-clone configuration failures so they cannot subsequently turn a successfully cloned Project into `error`. There must be no await that can route to the clone-failure handler after readiness is published.

Verification: Seed a pending task, make the post-clone manager setup raise, and schedule a dequeue when `wake()` is called. Assert the task remains pending whenever the final Project status is `error`; also assert a fully successful clone wakes and releases it.

Finding fingerprint: `532ec15d21eea640768fa15c5fde63ab80819a3416158922e296bf2bf076c16e`

File: backend/api/tasks.py [hunk 5]
```diff
@@ -4780,6 +4796,7 @@ async def _update_task_impl(
         if "project_id" in updates:
             # Project.local_path is the sole workspace authority.  Never carry
             # a caller-supplied or stale host path across a Project move.
+            _require_dispatchable_project(target_project)
```
Comment (youchengsong) [inline backend/api/tasks.py]: **[MEDIUM] Unchanged project IDs make unrelated task edits fail**

Reviewer: `senior_engineer` · Category: `regression`

Evidence: `_update_task_impl` calls `_require_dispatchable_project(target_project)` whenever `project_id` is present in the update payload, without checking whether it differs from the task's current project. Thus a full-form PUT carrying the existing project ID is rejected after that Project enters `error`, even if the user is only editing another field.

Impact: Existing queued tasks intentionally retained for re-clone recovery can no longer be edited by clients that submit the current project ID, despite the update not introducing a new association or attempting dispatch.

Required fix: Apply this admission check only when the effective project assignment changes to the error Project. Continue allowing edits that retain the existing project association; the dequeue gate already prevents execution.

Verification: Create a task on a ready Project, change the Project to `error`, then PUT a title change together with the unchanged `project_id` and assert success. Retain the existing test proving that moving a task from another Project into the error Project returns 422.

Finding fingerprint: `da961b0e2e70914b40155b0bc4072d2534fe79c896c51c82f8acd2f788926f22`

File: backend/services/instance_manager.py [hunk 6]
```diff
@@ -5270,13 +5270,29 @@ async def admit_claude_pty_transport() -> None:
...
         claude_delivery_git_boundary = None
         claude_task_runtime_scope_reserved = False
+        codex_task_runtime_scope_reserved = False
         if (
             provider == "claude"
             and task_id
             and not pr_review_task
         ):
             self._reserve_task_runtime_scope(task_id)
             claude_task_runtime_scope_reserved = True
```
Comment (youchengsong) [inline backend/services/instance_manager.py:5280]: **[HIGH] Codex scope reservation can remain leaked when launch fails before _launch_impl**

Reviewer: `qa_engineer` · Category: `concurrency`

Evidence: The new Codex path calls _reserve_task_runtime_scope(task_id) before Codex spec construction and process launch. The patch only describes discard handling in the later _launch_impl failure path; exceptions during the preceding setup/spec-building work can exit admit_claude_pty_transport without adopting or discarding this reservation.

Impact: A failed Codex launch can leave the task runtime scope permanently reserved, causing subsequent turns to be rejected or preventing cleanup of the task's runtime files. This is especially plausible for configuration/spec errors that occur before subprocess registration.

Required fix: Track the Codex reservation in the enclosing cleanup/finally path exactly like the Claude reservation, discarding it whenever no exact-generation owner was adopted; keep the _launch_impl cleanup as an additional path for failures after handoff.

Verification: Add a regression test that forces an exception after reserve but before process registration/spec launch and asserts the task runtime scope reservation is released and a subsequent turn can reserve the scope.

Finding fingerprint: `a653019c018ed93501dd4f9efd67b30311a64610f38c2e130c29a534270f4945`

File: backend/services/instance_manager.py [hunk 7]
```diff
@@ -5270,13 +5270,29 @@ async def admit_claude_pty_transport() -> None:
         claude_unrestricted_allowed_rules: tuple[str, ...] | None = None
```
Comment (youchengsong) [inline backend/services/instance_manager.py:5270]: **[MEDIUM] Codex reserve/adopt race fix has no concurrency regression test**

Reviewer: `qa_engineer` · Category: `testing`

Evidence: The PR adds a new reserve-to-exact-generation-adopt protocol for Codex app-server and exec launches, intended to prevent an older lifecycle finally block from deleting a newer turn's materialized MCP files. The supplied tests cover project readiness and clone behavior only; no test exercises overlapping generations, delayed spawn/adoption, or cleanup of an old generation.

Impact: The central safety claim remains unverified, and a future change can reintroduce deletion of a live turn's runtime scope without CI detecting it.

Required fix: Add deterministic tests for overlapping Codex generations: pause the newer launch after scope materialization and verify old lifecycle cleanup cannot remove it; cover both app-server adapter and exec-process adoption plus failed-launch cleanup.

Verification: Use controlled barriers/fakes around reserve, process registration, and lifecycle finally; assert exact-generation ownership and filesystem scope contents throughout the overlap.

Finding fingerprint: `64ede89c14de411b3e1ae8ede866d0375f3b957712274e7d0f702d0d485d8c5b`

**Comment by lvhaojie456** [comment]
已按 Reviewer Panel 的 7 条 finding 更新（新 head `e92f3330`），逐条回应：

**1 & 3. SSH clone 无 GIT_SSH_COMMAND 时仍可交互（principal + senior）— 已修复**
clone 作用域现在**始终**注入 batch-mode SSH 命令：有配置则增补 `-o BatchMode=yes`（保留既有选项），无配置用默认 `ssh -o BatchMode=yes`；未触碰共享 `_build_git_env`。测试按要求更新：`test_clone_no_config_no_env` 现在断言零配置下 `GIT_SSH_COMMAND == "ssh -o BatchMode=yes"` + `GIT_TERMINAL_PROMPT=0` + `stdin=DEVNULL`（并 `delenv` 隔离宿主继承）；`test_clone_passes_git_env_with_ssh` 断言既有命令被增补而非替换。

**2 & 4. wake 早于 clone 成功终局（principal + senior）— 已修复**
`ready` 提交现在是 clone 的最终发布：失败处理器之后 `return`，成功尾部移出 `try`。Delivery Monitor 自动配置在 ready 之后以独立 try/except 隔离（失败只记日志，Project 保持 ready——采纳 finding 给出的第二种修法，因为 `delivery_setup` 会跳过非 ready 项目，无法前置）；`_wake_dispatcher()` 挪到全部后置步骤之后。新增 `test_post_clone_setup_failure_cannot_reverse_ready`：强制自动配置抛异常，断言 Project 终态 ready、等待任务保持 pending 无注记、且事件顺序为 auto_config → wake。

**5. PUT 带未变更 project_id 误拒（senior）— 已修复**
门禁改为仅在 `updates["project_id"] != task.project_id` 时触发；保留原关联的全量表单编辑放行（出队门禁已阻止执行）。新增 `test_update_task_keeps_unchanged_error_project_editable`（项目转 error 后 PUT 改标题+原 project_id → 200），原 move-to-error 422 测试保留。

**6. Codex reservation 早期泄漏（qa，HIGH）— 经核实不成立，已补测试锁死**
`_launch_locked`（reserve 所在处）在整个代码库只有一个调用点：`_launch_impl` 的 3395 行，包在 `except BaseException`（3436）里；该分支的 discard（3479-3486）按 `task_id in self._task_runtime_scope_pending` 判断、provider 无关——spec 构建等 reserve 之后、注册之前的任何异常都会释放 reservation（唯一保留场景是 unresolved generation 的 fail-closed，与 Claude 语义一致且是既有设计）。`admit_claude_pty_transport` 是 `_launch_locked` 内部的回调闭包，不构成独立逃逸路径。按 verification 要求新增 `test_codex_launch_failure_after_reserve_discards_reservation`：在 reserve 之后、进程注册之前强制抛异常，断言 pending 已清、`cleanup_mcp_config` 恰好调用一次、下一代次可立即重新 reserve。

**7. 缺并发回归测试（qa）— 已补**
新增 `test_codex_scope_reservation_blocks_stale_lifecycle_cleanup`：覆盖「新代次已物化、未 spawn」窗口内旧 lifecycle 终态清理必须跳过（pending 栅栏）、adopt 交接后仍跳过、只有 exact 代次终态释放才清理。adoption 机制对 app-server `CodexTurnProcess` adapter 与 exec `Process` 是同一个按对象身份的 `_adopt_task_runtime_scope_direct`，两个调用点各自传入本 turn 的精确对象，测试以对象身份断言覆盖两者共享的协议。

**验证**（Linux 容器，python:3.12 + `--init`，干净 clone + patch）：readiness + git_credentials 50 passed；scope 回归 3 passed；task_queue/api_tasks/api_projects/api_project_todos/isolation 定向回归 485 passed / 2 skipped；instance_manager 全文件 530 passed。仅存 2 个失败（`test_delete_task_preserves_reaped_parent_with_live_generation`、`test_cloudrouter_claude_pty_projects_direct_auth_for_model_only`）经逐版本对比确认在上游 main `4cf021f7` 上以完全相同方式失败，为既有环境性问题，非本分支引入。

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Review by youchengsong — COMMENTED** [review]
# CCM reviewer panel: changes required

The panel found 1 open blocking finding on this reviewed PR head. Address the inline finding threads before merging.

## Reviewer summaries

### Principal engineer — Passed

The patch fits the existing queue, project, launch-isolation, and runtime-scope boundaries. It adds layered readiness enforcement, preserves exempt task creation paths, isolates post-clone optional failures after ready publication, and adds targeted regression coverage. No concrete blocking architectural or behavioral issue is evidenced by the supplied diff.

Finding count: none.

### Senior engineer — Passed

The supplied changes provide coherent project-readiness gating, noninteractive clone behavior, typed missing-directory failures, and Codex runtime-scope ownership handling. The reviewed paths include targeted regression coverage for the stated race and state-transition cases; no concrete blocking correctness, security, architecture, or maintainability issue was found.

Finding count: none.

### QA engineer — Changes required

The readiness gate is broadly covered, but its SQL predicate does not fail closed for NULL project statuses, allowing potentially unusable projects to dispatch.

Finding count: 1 total (1 blocking, 0 advisory).

Detailed evidence, impact, required fixes, and verification steps are published in dedicated finding threads/comments.

File: backend/services/task_queue.py [hunk 8]
```diff
@@ -169,6 +169,36 @@ def _dispatcher_scope_predicate():
...
+    without touching ``Task.status``.
+
+    Written as "no non-ready Project row exists" so Tasks without a Project
+    (manual ``target_repo``) and Tasks whose Project row was deleted keep the
+    existing dispatch behavior.
+    """
+
+    from backend.models.project import Project
+
+    unready_project = (
```
Comment (youchengsong) [inline backend/services/task_queue.py:193]: **[MEDIUM] NULL project status bypasses the readiness gate**

Reviewer: `qa_engineer` · Category: `correctness`

Evidence: The correlated subquery filters with `Project.status != "ready"`. In SQL, that expression evaluates to UNKNOWN for a NULL status, so no unready row is found and `~unready_project` allows the Task through. The predicate therefore treats NULL as ready even though only the literal `ready` is accepted by the intended contract.

Impact: Legacy or malformed Project rows with a NULL status can be dequeued and launched using a missing or invalid checkout, recreating the bare ENOENT/retry-budget failure this patch is intended to prevent.

Required fix: Make the predicate fail closed by explicitly treating NULL as non-ready (for example, use `Project.status.is_(None) | (Project.status != "ready")` in the subquery), or enforce/backfill a NOT NULL status invariant before relying on this gate.

Verification: Add a queue test creating a Project with `status=None` and assert its pending Task remains pending until the status is changed to `ready`.

Finding fingerprint: `0af1dc5675f9cebee8caab3015e7582ee84c0489205cc5e14fc8be03301b4ec7`

**Comment by lvhaojie456** [comment]
第二轮唯一的 finding（NULL project status 绕过门禁）已在新 head `d116fca5` 修复：

- 谓词的 correlated-EXISTS 改为显式 `or_(Project.status.is_(None), Project.status != "ready")`——NULL（legacy/异常行）按 fail-closed 处理，任务保持排队直到状态被修复为 `ready`。
- 按 verification 要求新增 `test_dequeue_holds_task_with_null_project_status`：当前模型的 `status` 列是 NOT NULL，测试用 `to_metadata` 放宽 nullable 的 schema 副本重建 legacy 形状，并以 **Core UPDATE** 写入真实 NULL（注意 ORM 对 `status=None` 会应用列默认值 `'pending'`，直接构造会得到假绿测试）。红绿已验证：移除 guard 时该测试确实失败（任务被派发）。
- Linux 容器复验：`test_project_readiness.py` + `test_task_queue.py` 全文件 134 passed；唯一失败仍是上游 main 上以相同方式失败的既有环境性用例（`test_delete_task_preserves_reaped_parent_with_live_generation`）。

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Review by youchengsong — COMMENTED** [review]
# CCM reviewer panel: changes required

The panel found 1 open blocking finding on this reviewed PR head. Address the inline finding threads before merging.

## Reviewer summaries

### Principal engineer — Changes required

The readiness and launch guards address the reported failure chain, but clone-failure note synchronization can overwrite unrelated task diagnostics.

Finding count: 1 total (1 blocking, 0 advisory).

### Senior engineer — Passed

The patch provides consistent project-readiness gating, fail-closed queue behavior, actionable working-directory errors, noninteractive clone handling, and runtime-scope ownership protections. The supplied changes and regression tests show no concrete blocking correctness, security, architecture, concurrency, or maintainability issue.

Finding count: none.

### QA engineer — Passed

The patch delivers the claimed project-readiness gates, non-interactive clone behavior, missing-working-directory errors, runtime-scope handling, and targeted regression coverage. No concrete blocking QA finding is evident in the supplied diff.

Finding count: none.

Detailed evidence, impact, required fixes, and verification steps are published in dedicated finding threads/comments.

File: backend/api/projects.py [hunk 9]
```diff
@@ -915,6 +917,80 @@ async def _prepare_existing_project_remote(
...
+    """Best-effort dispatcher nudge after a Project flips to ready.
+
+    Swallow every failure: raising here would land in the clone coroutine's
+    ``except`` block and mislabel an already-committed successful clone as an
+    error. The dispatcher's 2s poll remains the fallback.
+    """
+    try:
+        from backend.main import dispatcher
+
+        if dispatcher is not None:
```
Comment (youchengsong) [inline backend/api/projects.py:953]: **[MEDIUM] Do not overwrite existing task errors when annotating clone failures**

Reviewer: `principal_engineer` · Category: `correctness`

Evidence: `_sync_waiting_task_clone_notes` filters pending tasks only by `project_id`, `worker_id IS NULL`, and `shared_from_id IS NULL`; when `reason` is non-null it unconditionally sets `Task.error_message` to the project clone note. It does not require the field to be empty or already contain this helper's prefix.

Impact: A pending task that already has an independent launch or validation error can lose that diagnostic when the project clone fails. Subsequent UI and retry handling will report the project clone failure instead of the task-specific cause, and the original error is not recoverable from the task row.

Required fix: Restrict failure annotation to tasks with no existing `error_message` (or to notes previously generated by this helper), or preserve/compose the existing task error without replacing it. Keep the existing prefix-only rule for clearing notes.

Verification: Create a pending task for a project with `error_message='independent task error'`, invoke `_sync_waiting_task_clone_notes(project_id, 'clone failure')`, and assert the independent message is preserved; separately verify an empty task receives the clone note and that a subsequent success clears only the generated note.

Finding fingerprint: `36bbf8aaa937d56d93a27f0ab56f901ef55aa63925fd52a8820f73a1defa3965`

**Comment by lvhaojie456** [comment]
第三轮唯一的 finding（clone 注记覆盖任务独立诊断）已在新 head `a9441f1d` 修复：

- `_sync_waiting_task_clone_notes` 写入侧补上与清除侧对称的守卫：失败注记只写入 `error_message` 为 NULL/空串、或已带本 helper 前缀（`Project clone failed: `）的 pending 任务；任务已有的独立启动/校验诊断原样保留，不再被 clone 注记替换。清除侧 prefix-only 规则保持不变。
- 按 verification 要求新增 `test_clone_note_annotation_preserves_foreign_error_messages`，覆盖全部三个断言场景：① 已有独立错误（`independent task error`）的任务注记后原文案保留；② 空任务收到 clone 注记；③ 已带旧 helper 注记的任务可更新为新失败原因；随后 clone 成功清除时只清 helper 生成的注记，独立错误仍在。
- 红绿已验证：撤销 projects.py 守卫时该测试确实失败（独立错误被覆盖）。
- 回归：`test_project_readiness.py` 17 passed；连同 `test_task_queue.py` / `test_git_credentials.py` / `test_api_projects.py` 共 203 passed / 1 failed——唯一失败 `test_delete_task_preserves_reaped_parent_with_live_generation` 已在基线 `4cf021f7`（main）上复验以相同方式失败（既有环境性用例），零回归。
- TEST.md 用例表与 PROGRESS.md 复盘已同步（教训：成对的写入/清除逻辑必须共享同一识别谓词，只守一侧等于没守）。

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Review by youchengsong — COMMENTED** [review]
CCM automated review passed. The exact reviewed head `a9441f1d05a8b7ca5bc293d52e5cae767140f66f` is ready to merge.
